Counterexamples/PolynomialIsDomain.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2025 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.Algebra.GroupWithZero.TransferInstance
+import Mathlib.Algebra.Order.Ring.Nat
+import Mathlib.Algebra.Ring.Equiv
+import Mathlib.Algebra.Polynomial.Degree.Operations
+
+/-!
+# A commutative semiring that is a domain whose polynomial semiring is not a domain
+
+`NatMaxAdd` is the natural numbers equipped with the usual multiplication but with maximum as
+addition. Under these operations it is a commutative semiring that is a domain.
+However, in the polynomial semiring `NatMaxAdd[X]`, we have the identity
+`(X + 1) * (X ^ 2 + 1) = (X + 1) * (X ^ 2 + X + 1)`, so `NatMaxAdd[X]` is not a domain.
+This shows `Polynomial.instIsDomain` cannot be generalized from to semirings whose
+addition is not cancellative.
```

Comment:
TODO: if `a + b = a + c`, then `(X + 1) * (a * X ^ 2 + b * X + a) = (X + 1) * (a * X ^ 2 + c * X + a)`; this provides a counterexample for every commutative semiring whose addition isn't cancellative.